src/lib/libast/man/ftwalk.3
.TH FTWALK 3
.SH NAME
\fBftwalk\fR \- file tree walker
.SH SYNOPSIS
.ta .75i 1.5i 2.25i 3i 3.75i 4.5i 5.25i 6i
.PP
.nf
.ft CR
#include <ftwalk.h>

int ftwalk(char* path, int (*userf)(struct FTW* ftw), int options,
           int (*comparf)(struct FTW* ftw1, struct FTW* ftw2));
int ftwflags(void);
.ft
.fi
.SH DESCRIPTION
.PP
\fIFtwalk\fR traverses a directory hierarchy using depth-first search.
Upon visiting each file or directory in the hierarchy, it calls
the user function \fIuserf\fP to process that file or directory.
On a directory object, \fIuserf\fR may be called twice, once in preorder
and once in postorder.
On a terminal object such as a file or an unreadable directory,
\fIuserf\fP is called only once.
Cycles due to hard links or symbolic links are detected
to avoid infinite loops.
.PP
\fIPath\fR is the starting point of the search.
It may be an absolute path name or a path name relative to
the current directory.
If \fIpath\fR is a null pointer or points to an empty string, it is treated
as if it points to the current (dot) directory.
.PP
\fIOptions\fR consists of zero or more of the following bits:
.IP
FTW_CHILDREN:
This implies preorder calls to \fIuserf\fR on directory objects.
On such a call to \fIuserf\fR,
the field \fIftw->link\fR (below) points to a link list of
the children of the respective directory.
Upon returning from \fIuserf\fP,
if the field \fIftw->status\fR of any child object
is set to FTW_SKIP (below), that child is pruned from the search.
.IP
FTW_DELAY: When \fBFTW_CHILDREN\fP is turned on,
the fields \fIftw->statb\fP (\fIstruct stat\fP) of children objects
remain undefined until these objects are visited.
.IP
FTW_DOT: Do not use \fIchdir\fR(2) during the traversal.
Normally \fIchdir\fR is used so that
the base name of the object about to be processed can be used
in accessing its data.
This can enhance \fIftwalk\fR efficiency but certain program effects
such as core dumps may be generated in unexpected places
or may not even be generated at all.
Whenever \fIchdir\fR generates an error, if possible,
the current directory is restored to the starting directory
(see FTW_NAME and FTW_PATH).
.IP
FTW_MULTIPLE: The \fIpath\fP argument is treated as a \fIchar**\fP
pointer to a null-terminated array of path names.
All hierarchies rooted at these paths will be searched
.IP
FTW_POST: Calls to the user function are issued only in postorder.
That is, \fIuserf\fP is called on a directory only after its descendants have
been processed.
The absence of this bit indicates that calls to the user functions
are issued in preorder. That is, \fIuserf\fP is
called on a directory before its descendants are processed.
.IP
FTW_PHYSICAL: Use \fIlstat\fR(2) instead of \fIstat\fR(2) to get
file status and allow detection of symbolic links.
In addition, if each component
of the absolute path to the starting object has search permission,
the absolute path is used for early detection of cycles.
.IP
FTW_META|FTW_PHYSICAL: Use \fIstat\fR(2) for top level file status and
\fIlstat\fR(2) for all other files.
Used to implement the POSIX
.B \-H
option for commands like
.IR ls .
.IP
FTW_TWICE: Calls to the user function are issued in both preorder and postorder
for directory objects.
.IP
FTW_USER: The first of 6 user defined option bits.
These bits are ignored by \fIftwalk\fP.
.PP
\fIUserf\fR is a user supplied function that is
called upon different visits of an object.
If the return value of \fIuserf\fR is non-zero,
\fIftwalk\fR returns immediately with the same value.
The \fIuserf\fP prototype is:
.PP
.nf
	int userf(struct FTW* ftw)
.fi
.PP
\fBstruct FTW\fP contains at least the following elements:
.PP
.nf
    struct FTW*    link;    /* link list of children */
    struct FTW*    parent;  /* parent object on the search path */
    union
    {
    long           number;  /* local number */
    void*          pointer; /* local pointer */
    }              local;   /* user defined */
    struct stat    statb;   /* stat buffer of this object */
    char*          path;    /* full pathname */
    short          pathlen; /* strlen(path) */
    unsigned short info;    /* type of object */
    unsigned short status;  /* status of object */
    short          level;   /* depth of object on the search path */
    short          namelen; /* strlen(name) */
    char           name[];  /* file name of object */
.fi
.PP
The \fIlink\fR field is normally NULL.
If the option FTW_CHILDREN was turned on,
it points to the start of the list of children
of the directory being visited in preorder.
Finally, if the directory being visited causes a cycle,
\fIlink\fR points to the object on the search path that is
identical to this directory. Note that if FTW_PHYSICAL was turned on,
this may point to a directory that is an ancestor of the starting
object.
.PP
The \fIparent\fR field points to the parent object
on the search path. For convenience, a parent object is also supplied for
the starting object.
In this case, except for the \fIlocal\fR field which is initialized
to 0 and the \fIlevel\fR field which contains a negative number,
the rest of the structure may be undefined.
.PP
The \fIinfo\fR field indicates the type of the object
being visited and the type of the visit. The types are:
.IP
FTW_D: A directory being visited in preorder, i.e.,
none of its children has been visited by the search.
.IP
FTW_DNX: A directory being visited in preorder that does not have
search permission.
.IP
FTW_DP: A directory being visited in postorder, i.e., all of its
descendants have been completely processed.
.IP
FTW_DC: A directory that causes cycles. This is a terminal object.
.IP
FTW_DNR: A directory that cannot be opened for reading.
This is a terminal object.
.IP
FTW_F: An ordinary file.
.IP
FTW_SL: A symbolic link.
Unless FTW_FOLLOW (below) is issued by the user function,
this object is terminal.
.IP
FTW_NS: \fIStat\fR failed on this object.
The stat buffer \fIstatb\fR is undefined.
This object is terminal.
.PP
The \fIstatus\fR field of \fIstruct FTW\fR is used to communicate information
between \fIftwalk\fR and \fIuserf\fR. On calls to \fIuserf\fR, it has one of
two values:
.IP
FTW_NAME: The name of the object as defined in \fIftw->name\fR
should be used for accessing its file information.
This is because \fIchdir\fR(2) has been used
to set the current directory to a suitable place (see FTW_CHDIR).
.IP
FTW_PATH: The argument \fIpath\fR of \fIuserf\fR should be used
for accessing the file information of the object.
.PP
Upon returning, \fIuserf\fR may set the \fIstatus\fR field
to one of the following values:
.IP
FTW_AGAIN: If this is a directory object being visited in postorder,
it will be processed \fIagain\fR as if it had not been visited.
.IP
FTW_NOPOST: If this is a directory object being visited in preorder,
the user function will not be called on its postorder visit.
.IP
FTW_SKIP: This object and its descendants are pruned from the search.
.IP
FTW_FOLLOW: If this object is a symbolic link,
follow the link to its physical counterpart.
.PP
\fIComparf\fR, if not NULL, is a pointer to a function
used to define a search ordering for children of a directory.
If FTW_CHILDREN is turned on, the ordering of the children of
a directory is done before the preorder call to \fIuserf\fR on that directory.
Therefore, in that case, \fIftw->link\fR will point to the smallest child.
.PP
The \fIcomparf\fP prototype is:
.PP
.nf
	int comparf(struct FTW* ftw1, struct FTW* ftw2)
.fi
.PP
\fIComparf\fR should return a value <0, 0, or >0 to indicate whether
\fIftw1\fR is considered smaller, equal, or larger than \fIftw2\fR.
.PP
\fIFtwalk\fR normally returns 0.
On hard errors such as running out of memory, it returns -1.
\fIFtwalk\fR may also return other values as discussed with respect
to \fIuserf\fR.
.PP
\fIFtwflags\fR returns a combination of \fB0, FTW_META, FTW_PHYSICAL\fR
according to the
preferences specified by
\fBastconf("PATH_RESOLVE",0,0)\fR:
.TP
.B logical
0
.TP
.B metaphysical
.B "FTW_META|FTW_PHYSICAL"
.TP
.B physical
.B FTW_PHYSICAL
.SH HISTORY
\fIFtwalk\fR performs similar functions as that of
the routine \fIftw\fR provided in System V.
However, it is more general than \fIftw\fR
and suitable for use as a base in implementing
popular tools such as \fIls\fR, \fIfind\fR, \fItar\fR, \fIdu\fR, and \fIrm\fR.
\fIFtwalk\fR also handles symbolic links and hard links gracefully.
.SH AUTHORS
Phong Vo, Glenn Fowler, Dave Korn
.SH SEE ALSO
.IR find (1),
.IR rm (1),
.IR du (1),
.IR ls (1),
.IR tar (1),
.IR stat (2),
.IR symlink (2),
.IR fts (3),
.IR nftw (3),
.IR path (3)


NEWS
This documents significant changes in the dev branch of ksh 93u+m.
For full details, see the git log at: https://github.com/ksh93/ksh
Uppercase BUG_* IDs are shell bug IDs as used by the Modernish shell library.

2026-09-23:

- Fixed failures (assignment data loss) when a nested associative array was
  assigned to a name reference that pointed to a variable in a calling
  function's local scope.

- The sleep builtin will now reject values beyond UINT_MAX and negative numbers.

- Fixed bugs that caused many builtins (such as kill) to mishandle out of
  range numbers.

2026-09-21:

- Fixed a bug in nested indexed array assignments where values that happened
  to be the names of declaration commands would be executed as those
  declaration commands. For example, arr=((a b c)(typeset -p PATH)) now
  works correctly by assigning 'typeset', '-p' and 'PATH' as array values to
  arr[1][0], arr[1][1] and arr[1][2] instead of executing 'typeset -p PATH'.

- Fixed a buffer overflow that could occur when creating a type with
  'typeset -T', creating a function, then attempting to use 'typeset -p'.

2026-09-19:

- Fixed a longstanding bug where already-assigned associative array elements
  cannot be used in default assignment values of another element of same
  array -- for example, ${x[1]:=${x[0]}} failed to work correctly.

- Fixed default assignments on unset associative array elements, e.g.
  ${x[sub]=value} or ${x[sub]:=value}, which failed to assign the value to
  the element and instead left it empty or nonexistent.

- Fixed a bug in unsetting an associative array where the values of the array
  would be cleared but the former array variable still had leftover attributes.

2026-09-17:

- Fixed multiple memory leaks occurring upon subshell exit when assigning a
  value to a variable with a shell discipline function within a subshell.

- Fixed a crash occurring when assigning, within a subshell, an invalid
  value to a variable of a type defined by the 'enum' command.

2026-09-16:

- Fixed: when the special IFS (field splitting characters) variable was
  assigned to within a ${ shared-state command substitution }, the state of IFS
  failed to persist beyond it, and a use after free (potential crash) occurred.

- Fixed bounds checking when assigning to fixed-size arrays (declared like,
  e.g., 'typeset -a array[4]'). Assigning to an out-of-range subscript now
  produces the expected error instead of potentially crashing.

2026-09-15:

- Fixed a memory leak when assigning, within a subshell, a value to a
  special variable (one of IFS, FPATH, CDPATH, SHELL, ENV, VISUAL, EDITOR,
  HISTFILE, OPTIND, SECONDS, RANDOM,.sh.match, .sh.version, LINENO, _,
  LC_CTYPE, LC_NUMERIC, LC_COLLATE, LC_MESSAGES, LC_ALL, and LANG).

2026-09-13:

- Fixed a memory leak that occurred when converting an existing scalar
  variable to an associative array using 'typeset -A'.

2026-09-12:

- Implemented advisory locking for $HISTFILE (~/.sh_history by default) so
  that multiple ksh instances can write to the history file without risking
  data loss when one of them decides to trim the file. Note that ksh
  versions before 2026-09-12 will ignore the lock, so the risk continues to
  exist if an older ksh binary shares write access to the file.

- History file trimming no longer requires write access to the directory in
  which the history file resides and will no longer change the directory's
  last-modified timestamp. However, it now requires a usable /tmp or $TMPDIR.

- I/O errors while writing to the history file are no longer silently ignored.

- Fixed a possible crash when attempting to set an integer variable to a
  multibyte character.

2026-09-10:

- Fixed a rare crash that could occur if a signal (such as interrupt, Ctrl+C)
  occurred while cleaning up the state of a virtual subshell upon its exit.

2026-09-08:

- Fixed: PWD and ENV were not properly reinitialised when executing a shell
  script without a #! path.

- Fixed a possible crash after unsetting RANDOM. Bug introduced on 2022-07-21.

2026-09-04:

- Fixed a longstanding bug in which output data loss or corruption could
  intermittently occur if, within a command substitution, standard error is
  redirected to standard output and, within that redirection, standard
  output is then connected to a pipe.

2026-09-02:

- Fixed a bug, introduced on 2025-05-30, that made 'print -s' ineffective
  at writing to the history file. This functionality is now restored.

2026-08-31:

- Fixed a crash in the expr(1) path-bound built-in command that could occur
  when using the non-standard 'match' or 'index' keyword.

2026-08-30:

- Fixed a bug that caused non-importable environment variables (i.e., those
  with names that are not valid shell variable names) to disappear from the
  environment of child processes after invoking more than two external
  commands, potentially crashing the shell. Bug introduced on 2025-05-30.

2026-08-29:

- If an error (such as a syntax error) occurs in a trap action, the error
  message will now have the word "(trap)" in the call stack for clarity.

2026-08-26:

- Following up to 2026-08-16, fixed a remaining problem in the 'read'
  built-in with combined -d and -n options when reading from a pipe; the
  custom delimiter is now correctly used instead of newline.

2026-08-16:

- Fixed a long-standing bug in the 'read' built-in: the -d option was
  ignored when combined with -n. The two options can now be used together.

2026-08-14:

- In the --man/--help output for built-in commands, text marked as italic
  is no longer rendered as bold + underlined, but only as underlined.

2026-08-13:

- Fixed corruption of $_, crashing the shell upon 'typeset -p _', after an
  error occurs trying to redefine a type defined with 'typeset -T'.

2026-08-06:

- Fixed ksh crashing upon parsing a very long list of simple commands
  separated by semicolons or, in 'eval' or dot script contexts, by newlines.


src/lib/libast/misc/ftwalk.c
/***********************************************************************
*                                                                      *
*               This software is part of the ast package               *
*          Copyright (c) 1985-2011 AT&T Intellectual Property          *
*          Copyright (c) 2020-2026 Contributors to ksh 93u+m           *
*                      and is licensed under the                       *
*                 Eclipse Public License, Version 2.0                  *
*                                                                      *
*                A copy of the License is available at                 *
*      https://www.eclipse.org/org/documents/epl-2.0/EPL-2.0.html      *
*         (with md5 checksum 84283fa8859daf213bdda5a9f8d1be1d)         *
*                                                                      *
*                 Glenn Fowler <gsf@research.att.com>                  *
*                  David Korn <dgk@research.att.com>                   *
*                   Phong Vo <kpv@research.att.com>                    *
*                  Martijn Dekker <martijn@inlv.org>                   *
*            Johnothan King <johnothanking@protonmail.com>             *
*                                                                      *
***********************************************************************/
/*
 * Glenn Fowler
 * AT&T Research
 *
 * ftwalk on top of fts
 */

#include <ast.h>
#include <ftwalk.h>

static struct
{
	int	(*comparf)(Ftw_t*, Ftw_t*);
} state;

/*
 * why does fts take FTSENT** instead of FTSENT*
 */

static int
ftscompare(Ftw_t* const* pf1, Ftw_t* const* pf2)
{
	return (*state.comparf)(*pf1, *pf2);
}

/*
 * the real thing -- well it used to be
 */

int
ftwalk(const char* path, int (*userf)(Ftw_t*), int flags, int (*comparf)(Ftw_t*, Ftw_t*))
{
	FTS*		f;
	FTSENT*		e;
	int		children;
	int		rv;
	unsigned short	oi;
	short		os;
	size_t		ns;
	size_t		nd;
	FTSENT*		x;
	FTSENT*		dd[2];

	flags ^= FTS_ONEPATH;
	if (flags & FTW_TWICE)
		flags &= ~(FTS_NOPREORDER|FTS_NOPOSTORDER);
	else if (flags & FTW_POST)
		flags |= FTS_NOPREORDER;
	else
		flags |= FTS_NOPOSTORDER;
	if (children = flags & FTW_CHILDREN)
		flags |= FTS_SEEDOT;
	state.comparf = comparf;
	if (!(f = fts_open((char* const*)path, flags, comparf ? ftscompare : 0)))
	{
		if (!path || !(flags & FTS_ONEPATH) && !(path = (const char*)(*((char**)path))))
			return -1;
		ns = strlen(path) + 1;
		if (!(e = newof(0, FTSENT, 1, ns)))
			return -1;
		e->fts_accpath = e->fts_name = e->fts_path = strcpy((char*)(e + 1), path);
		e->fts_namelen = e->fts_pathlen = ns;
		e->fts_info = FTS_NS;
		e->parent = e;
		e->parent->link = e;
		rv = (*userf)((Ftw_t*)e);
		free(e);
		return rv;
	}
	rv = 0;
	if (children && (e = fts_children(f, 0)))
	{
		nd = 0;
		for (x = e; x; x = x->link)
			if (x->info & FTS_DD)
			{
				x->statb = *x->fts_statp;
				x->info &= ~FTS_DD;
				dd[nd++] = x;
				if (nd >= elementsof(dd))
					break;
			}
		e->parent->link = e;
		rv = (*userf)((Ftw_t*)e->parent);
		e->parent->link = 0;
		while (nd > 0)
			dd[--nd]->info |= FTS_DD;
		for (x = e; x; x = x->link)
			if (!(x->info & FTS_D))
				x->status = FTS_SKIP;
	}
	while (!rv && (e = fts_read(f)))
	{
		oi = e->info;
		os = e->status;
		ns = e->status = e->path == e->fts_accpath ? FTW_PATH : FTW_NAME;
		nd = 0;
		switch (e->info)
		{
		case FTS_D:
		case FTS_DNX:
			if (children)
				for (x = fts_children(f, 0); x; x = x->link)
					if (x->info & FTS_DD)
					{
						x->statb = *x->fts_statp;
						x->info &= ~FTS_DD;
						dd[nd++] = x;
						if (nd >= elementsof(dd))
							break;
					}
			break;
		case FTS_DOT:
			continue;
		case FTS_ERR:
			e->info = FTS_NS;
			break;
		case FTS_NSOK:
			e->info = FTS_NSOK;
			break;
		case FTS_SLNONE:
			e->info = FTS_SL;
			break;
		}
		rv = (*userf)((Ftw_t*)e);
		e->info = oi;
		if (e->status == (ssize_t)ns)
			e->status = os;
		while (nd > 0)
			dd[--nd]->info |= FTS_DD;
	}
	fts_close(f);
	return rv;
}
